File: backend_python/app/services/connection_service.py
from datetime import datetime, timezone
from bson import ObjectId
from app.config.database import get_db
from app.models.common import serialize_doc


async def invite_assistant(doctor_id: str, clinic_id: str, assistant_phone: str) -> dict:
    db = get_db()
    assistant = await db.assistants.find_one({"phone": assistant_phone})
    if not assistant:
        raise ValueError("Assistant not found with that phone number")

    assistant_id = str(assistant["_id"])
    existing = await db.connection_requests.find_one({
        "doctor_id": doctor_id,
        "assistant_id": assistant_id,
        "status": "pending",
    })
    if existing:
        raise ValueError("Pending request already exists")

    doc = {
        "clinic_id": clinic_id,
        "doctor_id": doctor_id,
        "assistant_id": assistant_id,
        "initiated_by": "doctor",
        "status": "pending",
        "created_at": datetime.now(timezone.utc),
        "updated_at": datetime.now(timezone.utc),
    }
    result = await db.connection_requests.insert_one(doc)
    req = await db.connection_requests.find_one({"_id": result.inserted_id})
    return serialize_doc(req)


async def request_join(assistant_id: str, doctor_code: str) -> dict:
    db = get_db()
    doctor = await db.doctors.find_one({"doctor_code": doctor_code})
    if not doctor:
        raise ValueError("Doctor not found with that code")

    doctor_id = str(doctor["_id"])
    existing = await db.connection_requests.find_one({
        "doctor_id": doctor_id,
        "assistant_id": assistant_id,
        "status": "pending",
    })
    if existing:
        raise ValueError("Pending request already exists")

    doc = {
        "clinic_id": doctor.get("clinic_id"),
        "doctor_id": doctor_id,
        "assistant_id": assistant_id,
        "initiated_by": "assistant",
        "status": "pending",
        "created_at": datetime.now(timezone.utc),
        "updated_at": datetime.now(timezone.utc),
    }
    result = await db.connection_requests.insert_one(doc)
    req = await db.connection_requests.find_one({"_id": result.inserted_id})
    return serialize_doc(req)


async def accept_request(request_id: str, user_id: str, role: str) -> dict:
    db = get_db()
    req = await db.connection_requests.find_one({"_id": ObjectId(request_id), "status": "pending"})
    if not req:
        raise ValueError("Request not found or not pending")

    if role == "doctor" and req["doctor_id"] != user_id:
        raise ValueError("Not authorized")
    if role == "assistant" and req["assistant_id"] != user_id:
        raise ValueError("Not authorized")

    await db.connection_requests.update_one(
        {"_id": ObjectId(request_id)},
        {"$set": {"status": "accepted", "updated_at": datetime.now(timezone.utc)}}
    )
    await db.assistants.update_one(
        {"_id": ObjectId(req["assistant_id"])},
        {"$set": {"clinic_id": req.get("clinic_id"), "updated_at": datetime.now(timezone.utc)}}
    )

    # Clinic now has at least one assistant → solo_mode off.
    if req.get("clinic_id"):
        await db.clinics.update_one(
            {"_id": ObjectId(req["clinic_id"])},
            {"$set": {"solo_mode": False, "updated_at": datetime.now(timezone.utc)}},
        )

    req = await db.connection_requests.find_one({"_id": ObjectId(request_id)})
    return serialize_doc(req)


async def reject_request(request_id: str, user_id: str, role: str) -> dict:
    db = get_db()
    req = await db.connection_requests.find_one({"_id": ObjectId(request_id), "status": "pending"})
    if not req:
        raise ValueError("Request not found or not pending")

    if role == "doctor" and req["doctor_id"] != user_id:
        raise ValueError("Not authorized")
    if role == "assistant" and req["assistant_id"] != user_id:
        raise ValueError("Not authorized")

    await db.connection_requests.update_one(
        {"_id": ObjectId(request_id)},
        {"$set": {"status": "rejected", "updated_at": datetime.now(timezone.utc)}}
    )
    req = await db.connection_requests.find_one({"_id": ObjectId(request_id)})
    return serialize_doc(req)


async def get_pending(user_id: str, role: str) -> list:
    db = get_db()
    query = {"status": "pending"}
    if role == "doctor":
        query["doctor_id"] = user_id
    else:
        query["assistant_id"] = user_id

    result = []
    async for r in db.connection_requests.find(query):
        doc = serialize_doc(r)

        # Attach human-readable names so the frontend can display them
        try:
            doctor = await db.doctors.find_one({"_id": ObjectId(doc["doctor_id"])})
            doc["doctor_name"] = doctor.get("name") if doctor else None
        except Exception:
            doc["doctor_name"] = None

        try:
            assistant = await db.assistants.find_one({"_id": ObjectId(doc["assistant_id"])})
            if assistant:
                doc["assistant_name"] = assistant.get("name")
                doc["assistant_phone"] = assistant.get("phone")
                doc["qualification"] = assistant.get("specialty")
                doc["experience_years"] = assistant.get("experience_years")
                doc["city"] = assistant.get("city")
                doc["assistant_address"] = assistant.get("address")
            else:
                doc["assistant_name"] = None
                doc["assistant_phone"] = None
        except Exception:
            doc["assistant_name"] = None
            doc["assistant_phone"] = None

        # Attach clinic name
        try:
            if doc.get("clinic_id"):
                clinic = await db.clinics.find_one({"_id": ObjectId(doc["clinic_id"])})
                doc["clinic_name"] = clinic.get("name") if clinic else None
            else:
                doc["clinic_name"] = None
        except Exception:
            doc["clinic_name"] = None

        result.append(doc)
    return result


async def get_team(clinic_id: str) -> list:
    db = get_db()
    team = []
    
    # Doctors
    cursor = db.doctors.find({
        "clinic_id": clinic_id,
        "is_active": True,
    })
    async for u in cursor:
        doc = serialize_doc(u)
        doc.pop("password_hash", None)
        doc.pop("otp_hash", None)
        team.append(doc)

    # Assistants
    cursor = db.assistants.find({
        "clinic_id": clinic_id,
        "is_active": True,
    })
    async for u in cursor:
        doc = serialize_doc(u)
        doc.pop("password_hash", None)
        doc.pop("otp_hash", None)
        team.append(doc)

    return team


async def disconnect_assistant(doctor_id: str, clinic_id: str, assistant_id: str):
    db = get_db()
    assistant = await db.assistants.find_one({"_id": ObjectId(assistant_id), "clinic_id": clinic_id})
    if not assistant:
        raise ValueError("Assistant not in your clinic")

    await db.assistants.update_one(
        {"_id": ObjectId(assistant_id)},
        {"$set": {"clinic_id": None, "updated_at": datetime.now(timezone.utc)}}
    )
    await db.connection_requests.update_many(
        {"doctor_id": doctor_id, "assistant_id": assistant_id},
        {"$set": {"status": "rejected", "updated_at": datetime.now(timezone.utc)}}
    )

    # If no assistants remain, flip the clinic back to solo_mode so the doctor
    # regains AddPatient / PatientSearch capability.
    remaining = await db.assistants.count_documents({
        "clinic_id": clinic_id, "is_active": True,
    })
    if remaining == 0:
        await db.clinics.update_one(
            {"_id": ObjectId(clinic_id)},
            {"$set": {"solo_mode": True, "updated_at": datetime.now(timezone.utc)}},
        )

File: frontend/src/services/connectionService.ts
import api from './api';
import { ConnectionRequest, TeamMember, ClinicListItem, DoctorListItem } from '../types/connection.types';

export async function inviteAssistant(assistantPhone: string): Promise<ConnectionRequest> {
  const response = await api.post('/connection/invite', { assistantPhone });
  return response.data.request;
}

export async function requestToJoin(doctorCode: string): Promise<ConnectionRequest> {
  const response = await api.post('/connection/request', { doctorCode });
  return response.data.request;
}

export async function acceptRequest(requestId: string): Promise<void> {
  await api.put(`/connection/${requestId}/accept`);
}

export async function rejectRequest(requestId: string): Promise<void> {
  await api.put(`/connection/${requestId}/reject`);
}

export async function getPendingRequests(): Promise<ConnectionRequest[]> {
  const response = await api.get('/connection/pending');
  const raw: any[] = response.data.requests ?? [];
  return raw.map(normalizeRequest);
}

export async function getTeamMembers(): Promise<TeamMember[]> {
  const response = await api.get('/connection/team');
  const raw: any[] = response.data.members ?? [];
  return raw.map(normalizeMember);
}

export async function disconnectAssistant(assistantId: string): Promise<void> {
  await api.delete(`/connection/team/${assistantId}`);
}

export async function listClinics(search?: string): Promise<ClinicListItem[]> {
  const params = search ? { search } : {};
  const response = await api.get('/clinic/list', { params });
  const raw: any[] = response.data.clinics ?? [];
  return raw.map(normalizeClinic);
}

export async function getDoctorsByClinic(clinicId: string): Promise<DoctorListItem[]> {
  const response = await api.get(`/clinic/${clinicId}/doctors`);
  const raw: any[] = response.data.doctors ?? [];
  return raw.map(normalizeDoctor);
}

// ── normalizers ──────────────────────────────────────────────────────────────

function normalizeClinic(r: any): ClinicListItem {
  return {
    id: r.id ?? r._id ?? '',
    name: r.name ?? '',
    address: r.address ?? '',
    phone: r.phone ?? '',
    doctorName: r.doctor_name ?? r.doctorName ?? '',
    doctorSpecialty: r.doctor_specialty ?? r.doctorSpecialty ?? '',
    ownerId: r.owner_id ?? r.ownerId ?? '',
  };
}

function normalizeDoctor(r: any): DoctorListItem {
  return {
    id: r.id ?? r._id ?? '',
    name: r.name ?? '',
    specialty: r.specialty ?? '',
    regNumber: r.reg_number ?? r.regNumber ?? '',
    doctorCode: r.doctor_code ?? r.doctorCode ?? '',
  };
}

function normalizeRequest(r: any): ConnectionRequest {
  return {
    id: r.id ?? r._id ?? '',
    doctorId: r.doctor_id ?? r.doctorId ?? '',
    assistantId: r.assistant_id ?? r.assistantId ?? '',
    initiatedBy: r.initiated_by ?? r.initiatedBy ?? 'assistant',
    status: r.status ?? 'pending',
    doctorName: r.doctor_name ?? r.doctorName,
    assistantName: r.assistant_name ?? r.assistantName,
    clinicName: r.clinic_name ?? r.clinicName,
    createdAt: r.created_at ?? r.createdAt ?? '',
    qualification: r.qualification,
    experienceYears: r.experience_years ?? r.experienceYears,
    city: r.city,
    assistantAddress: r.assistant_address ?? r.assistantAddress,
    assistantPhone: r.assistant_phone ?? r.assistantPhone,
  };
}

function normalizeMember(r: any): TeamMember {
  return {
    id: r.id ?? r._id ?? '',
    name: r.name ?? '',
    phone: r.phone ?? '',
    role: r.role ?? 'assistant',
    lastActiveAt: r.last_active_at ?? r.lastActiveAt,
    qualification: r.qualification,
    experienceYears: r.experience_years ?? r.experienceYears,
    profileAddress: r.profile_address ?? r.profileAddress,
    city: r.city,
    specialty: r.specialty,
    regNumber: r.reg_number ?? r.regNumber,
  };
}
